server/http/src/service.rs

```diff
@@ -16,14 +19,36 @@ pub type GraphQLServiceResponse =
 /// A Hyper Service that serves GraphQL over a POST / endpoint.
 #[derive(Debug)]
 pub struct GraphQLService {
-    schema: Arc<Mutex<Option<Schema>>>,
+    schemas: Arc<RwLock<BTreeMap<String, Schema>>>,
     query_sink: Sender<Query>,
 }
 
 impl GraphQLService {
     /// Creates a new GraphQL service.
-    pub fn new(schema: Arc<Mutex<Option<Schema>>>, query_sink: Sender<Query>) -> Self {
-        GraphQLService { schema, query_sink }
+    pub fn new(schemas: Arc<RwLock<BTreeMap<String, Schema>>>, query_sink: Sender<Query>) -> Self {
+        GraphQLService {
+            schemas,
+            query_sink,
+        }
+    }
+
+    fn index(&self) -> GraphQLServiceResponse {
+        let len = self.schemas.read().unwrap().len();
+        match self.schemas.read().unwrap().values().next().clone() {
+            // If there's only 1 schema, redirect to it.
+            Some(schema) if len == 1 => Box::new(future::ok(
+                Response::builder()
+                    .status(StatusCode::SEE_OTHER)
+                    .header(
+                        header::LOCATION,
+                        header::HeaderValue::from_str(&format!("/{}", schema.name))
```

Comment:
@timmclean there are URL injection risks here, because the graph name is inserted directly in the redirect the URL and there are no restrictions on a name and no percent encoding, so it could redirect to any route. Do you have an opinion on a good solution here?

Reply:
Ah, interesting. Is `schema.name` just any UTF-8 string the user provides? That's tough. Usual solutions here are: (1) use an ID instead (randomly generated with only friendly characters), (2) generate a "sanitized" version of the name that uses only a whitelist of characters, (3) both but the ID takes precedence.

I think I'd suggest (1), unless you want the name in there for better UX. StackOverflow and many news sites do (3) for SEO reasons. (2) can work, but then you need to make sure that no two schemas have the same sanitized name.
Also, seems like a good idea to follow Jannis's earlier suggestion with using a URL path library to get a path.join function

Reply:
Thanks for the advice here! For now this is a naive naming system, we will build a Graph Naming System on ethereum contracts to solve things like name collisions. Allowing the user to choose a name makes a more friendly UI. Would percent encoding/decoding the name be a decent implementation of (2)?

Reply:
So, straight-up percent encoding doesn't entirely solve the problem, because `..` percent-encodes to `..` (and even if you force it to be `%2e%2e`, that still has the "parent directory" semantics). The sanitization approach wouldn't actually use encoding: it involves completely removing any characters that are not in the whitelist.

I _think_ you can probably get away with using [the method Jannis suggested earlier](https://docs.rs/url/1.7.1/url/struct.PathSegmentsMut.html#method.push), if you check afterwards that a component was actually pushed (that method treats `.`, `..`, and their encoded equivalents as no-ops -- it should treat them as errors in my opinion).  It's just a bit riskier, because it depends on implementation details of the server. E.g. imagine a server that ignores NUL bytes. It looks like the string `%2e%2e%00` would be accepted by the `push` method as a valid path component, but once that URL reaches the server, the server would ignore `%00` and see only `%2e%2e`, which decodes to `..` again.  I think that behaviour would be against spec, but, you know how that often goes...

Reply:
Sounds good!